src/java/org/apache/cassandra/index/sai/utils/RangeIntersectionIterator.java

```diff
@@ -63,54 +61,58 @@ private RangeIntersectionIterator(Builder.Statistics statistics, List<RangeItera
         this.processedRanges = new ArrayList<>(ranges.size());
     }
 
+    /**
+     * Fetches the next available item from the iterator, such that the item is not lower than the given key.
+     * If no such items are available, returns null.
+     */
+    private PrimaryKey nextOrNull(RangeIterator iterator, PrimaryKey minKey) {
+        iterator.skipTo(minKey);
+        return iterator.hasNext() ? iterator.next() : null;
+    }
+
     protected PrimaryKey computeNext()
     {
-        processedRanges.clear();
-        RangeIterator head = null;
-        PrimaryKey candidate = null;
-        while (!ranges.isEmpty())
+        // Range iterator that has been advanced in the previous cycle of the outer loop.
+        // Initially there hasn't been the previous cycle, so set to null.
+        RangeIterator alreadyAvanced = null;
+
+        // The highest primary key seen on any range iterator so far.
+        // It can become null when we reach the end of the iterator.
+        PrimaryKey highestKey = getCurrent();
+
+        outer:
+        while (highestKey != null && highestKey.compareTo(getMaximum()) <= 0)
         {
-            RangeIterator range = ranges.remove(0);
-            if (!range.hasNext())
-                return endOfData();
-            if (range.getCurrent().compareTo(getMinimum()) < 0)
-            {
-                range.skipTo(getMinimum());
-            }
-            if (candidate == null)
-            {
-                candidate = range.hasNext() ? range.next() : null;
-                if (candidate == null || candidate.compareTo(getMaximum()) > 0)
-                    return endOfData();
-                head = range;
-            }
-            else
+            // Try advance all iterators to the highest key seen so far.
+            // Once this inner loop finishes normally, all iterators are guaranteed to be at the same value.
+            for (RangeIterator range: ranges)
             {
-                if (!isOverlapping(head, range) || (range.skipTo(candidate) == null))
-                {
-                    return endOfData();
-                }
-                int cmp = candidate.compareTo(range.getCurrent());
-                if (cmp == 0)
+                if (range != alreadyAvanced)
```

Comment:
Doesn't this mean that if there are more than 2 iterators that the first iterator could be processed twice?